RemoteAudioDestinationProxy base class is platform specific

kkinnunen-apple opened the pull request:
#### 3fac3faf1d5b167d9fbbd9bf4555c133718f5044
<pre>
RemoteAudioDestinationProxy base class is platform specific
<a href="https://bugs.webkit.org/show_bug.cgi?id=247802">https://bugs.webkit.org/show_bug.cgi?id=247802</a>
rdar://problem/102238646

Reviewed by Eric Carlson.

The ifdefing causes problems when passing the needed information to
RemoteAudioDestinationManager::CreateAudioDestination. Due to this,
the GPUP side instance is created with
CreateAudioDestination+AudioSamplesStorageChanged message, which
causes other problems.

Instead, extract the needed platform-agnostic logic from
AudioDestinationCocoa to AudioDestinationResampler.
Use AudioDestinationResampler as the base-class of
RemoteAudioDestinationProxy, as that is the way the data
is transferred currently.

This is preliminary work to be able to make CARingBuffer
as a platform-agnostic AudioRingBuffer, in order to pass
the ring buffer during the remote audio destination creation.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::sampleRate const):
(WebCore::AudioDestination::AudioDestination):
* Source/WebCore/platform/audio/PushPullFIFO.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::AudioDestinationCocoa):
(WebCore::AudioDestinationCocoa::render):
(WebCore::AudioDestinationCocoa::numberOfOutputChannels const): Deleted.
(WebCore::AudioDestinationCocoa::framesPerBuffer const): Deleted.
(WebCore::AudioDestinationCocoa::start): Deleted.
(WebCore::AudioDestinationCocoa::stop): Deleted.
(WebCore::AudioDestinationCocoa::setIsPlaying): Deleted.
(WebCore::AudioDestinationCocoa::getAudioStreamBasicDescription): Deleted.
(WebCore::assignAudioBuffersToBus): Deleted.
(WebCore::AudioDestinationCocoa::hasEnoughFrames const): Deleted.
(WebCore::AudioDestinationCocoa::renderOnRenderingTheadIfPlaying): Deleted.
(WebCore::AudioDestinationCocoa::renderOnRenderingThead): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.h:
(): Deleted.
* Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.h:
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.cpp:
(WebCore::MockAudioDestinationCocoa::tick):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateScrollingNodeForViewportConstrainedRole):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::create):
(WebKit::RemoteAudioDestinationProxy::RemoteAudioDestinationProxy):
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::startRendering):
(WebKit::RemoteAudioDestinationProxy::stopRendering):
(WebKit::RemoteAudioDestinationProxy::renderQuantum):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:

Canonical link: <a href="https://commits.webkit.org/256729@main">https://commits.webkit.org/256729@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3655419cd26dbae21eca13cc66a8bcb1edbb1017

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5940 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29763 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106205 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6145 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34675 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89057 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102915 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102357 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83285 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31553 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88301 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74472 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40403 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/38064 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21205 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4671 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4334 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1149 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40487 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/audio/AudioDestination.h:
```diff
@@ -75,13 +75,16 @@ class AudioDestination : public ThreadSafeRefCounted<AudioDestination, WTF::Dest
-    explicit AudioDestination(AudioIOCallback&);
+    explicit AudioDestination(AudioIOCallback&, float sampleRate);
+
+    const float m_sampleRate;
```
Tiny nit: looks like this could be private.

eric-carlson commented on Source/WebCore/platform/audio/AudioDestinationResampler.cpp:
```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2019 Apple Inc. All rights reserved.
```
2019?

eric-carlson commented on Source/WebCore/platform/audio/AudioDestinationResampler.cpp:
```diff
@@ -0,0 +1,148 @@
+    ASSERT(!isMainThread());
+    ASSERT(numberOfFrames <= fifoSize);
+    Locker locker { m_fifoLock };
+    return m_fifo.pull(m_outputBus.ptr(), numberOfFrames);
```
It is good to assert when asked to pull too many frames in a debug build, but what about in a release build? Should we RELEASE_ASSERT, or clamp `numberOfFrames` to  `fifoSize`?

eric-carlson commented on Source/WebCore/platform/audio/AudioDestinationResampler.h:
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2019 Apple Inc. All rights reserved.
```
2019?

eric-carlson commented on Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
```diff
@@ -148,110 +102,24 @@ void AudioDestinationCocoa::stopRendering(CompletionHandler<void(bool)>&& comple
+        ASSERT(buffers[i].mDataByteSize >= numberOfFrames * sizeof(float));
+        auto* memory = reinterpret_cast<float*>(buffers[i].mData);
+        m_outputBus->setChannelMemory(i, memory, numberOfFrames);
     }
```
Like above, should we `RELEASE_ASSERT` or clamp?

eric-carlson commented on Source/WebCore/platform/audio/cocoa/AudioUtilitiesCocoa.cpp:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2019 Apple Inc. All rights reserved.
```
2019?

eric-carlson commented on Source/WebCore/platform/audio/cocoa/AudioUtilitiesCocoa.h:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2019 Apple Inc. All rights reserved.
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
```diff
@@ -173,8 +174,24 @@ void RemoteAudioDestinationProxy::renderQuantum()
+        ASSERT(buffers[i].mDataByteSize >= numberOfFrames * sizeof(float));
+        auto* memory = reinterpret_cast<float*>(buffers[i].mData);
+        m_outputBus->setChannelMemory(i, memory, numberOfFrames);
+    }
```
`RELEASE_ASSERT` or clamp?

eric-carlson reviewed (approved):
This is a great cleanup and refactoring, thank you!

kkinnunen-apple committed:
RemoteAudioDestinationProxy base class is platform specific
https://bugs.webkit.org/show_bug.cgi?id=247802
rdar://problem/102238646

Reviewed by Eric Carlson.

The ifdefing causes problems when passing the needed information to
RemoteAudioDestinationManager::CreateAudioDestination. Due to this,
the GPUP side instance is created with
CreateAudioDestination+AudioSamplesStorageChanged message, which
causes other problems.

Instead, extract the needed platform-agnostic logic from
AudioDestinationCocoa to AudioDestinationResampler.
Use AudioDestinationResampler as the base-class of
RemoteAudioDestinationProxy, as that is the way the data
is transferred currently.

This is preliminary work to be able to make CARingBuffer
as a platform-agnostic AudioRingBuffer, in order to pass
the ring buffer during the remote audio destination creation.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::sampleRate const):
(WebCore::AudioDestination::AudioDestination):
* Source/WebCore/platform/audio/PushPullFIFO.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::AudioDestinationCocoa):
(WebCore::AudioDestinationCocoa::render):
(WebCore::AudioDestinationCocoa::numberOfOutputChannels const): Deleted.
(WebCore::AudioDestinationCocoa::framesPerBuffer const): Deleted.
(WebCore::AudioDestinationCocoa::start): Deleted.
(WebCore::AudioDestinationCocoa::stop): Deleted.
(WebCore::AudioDestinationCocoa::setIsPlaying): Deleted.
(WebCore::AudioDestinationCocoa::getAudioStreamBasicDescription): Deleted.
(WebCore::assignAudioBuffersToBus): Deleted.
(WebCore::AudioDestinationCocoa::hasEnoughFrames const): Deleted.
(WebCore::AudioDestinationCocoa::renderOnRenderingTheadIfPlaying): Deleted.
(WebCore::AudioDestinationCocoa::renderOnRenderingThead): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.h:
(): Deleted.
* Source/WebCore/platform/audio/gstreamer/AudioDestinationGStreamer.h:
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.cpp:
(WebCore::MockAudioDestinationCocoa::tick):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateScrollingNodeForViewportConstrainedRole):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::create):
(WebKit::RemoteAudioDestinationProxy::RemoteAudioDestinationProxy):
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::startRendering):
(WebKit::RemoteAudioDestinationProxy::stopRendering):
(WebKit::RemoteAudioDestinationProxy::renderQuantum):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:

Canonical link: https://commits.webkit.org/256729@main
